In [11]:
import numpy as np  # Used to make a matrix type
array = [10,5,16,9,17,12,14,13,8,25,15,30,12,16,9,14,25,30,19,16,24,17,18,29,11,16,14,21,11,33,10]; 
L = len(array)
count_mat = np.zeros([L-1,L-1])
k_mat = np.zeros([L-1,L-1],dtype=object)

for i in range(0,L-1):
    for j in range(0,L-1):
        k_mat[i,j] = []

count_mat.shape

count_mat[0,0]

# Go through the number of matrices - 1 diagonals (i.e., len(array)-2)

# Variables
#
# i = row index
# j = column index
# k = matrix parenthesis index
#
# d = diagonal index (i.e., chain length)
#

for d in range(1,L-1):
    for i in range(0,L-d-1):
        min_val = 100000000000000000000
        j = i+d
        for k in range(i,j):  # go through all matrix paren index for a row
            c_val = count_mat[i,k]+count_mat[k+1,j]+array[i-1]*array[k]*array[j]
            if c_val<min_val:
                min_val = c_val
                min_k = k_mat[k+1,j]+k_mat[i,k]+[k]
        k_mat[i,j] = min_k
        count_mat[i,j] = min_val
        
print("# of operations: "+repr(count_mat[0,L-2]))
print("")
print("Index to 'split' multiplication vector with paranthesese: "+repr(k_mat[0,L-2]))
print("")

# of operations: 42165.0

Index to 'split' multiplication vector with paranthesese: [2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 0, 1]



Given our split index, our optimal matrix parenthesization is:

$((x_1(x_2))(x_3(x_4(x_5(x_6(x_7(x_8(x_9(x_{10}(x_{11}(x_{12}(x_{13}(x_{14}(x_{15}(x_{16}(x_{17}(x_{18}(x_{19}(x_{20}(x_{21}(x_{22}(x_{23}(x_{24}(x_{25}(x_{26}(x_{27}(x_{28}(x_{29}(x_{30})))))))))))))))))))))))))))))$

The number of operations required for the multiplication is: 42165